language/compiler/ir-to-bytecode/syntax/src/syntax.rs

```diff
@@ -60,6 +60,31 @@ fn consume_token<'input>(
     Ok(())
 }
 
+fn parse_comma_list<'input, F, R>(
+    tokens: &mut Lexer<'input>,
+    list_end_token: Tok,
+    parse_list_item: F,
+    allow_trailing_comma: bool,
+) -> Result<Vec<R>, ParseError<usize, anyhow::Error>>
+where
+    F: Fn(&mut Lexer<'input>) -> Result<R, ParseError<usize, anyhow::Error>>,
+{
+    let mut v = vec![];
+    if tokens.peek() != list_end_token {
+        loop {
+            v.push(parse_list_item(tokens)?);
+            if tokens.peek() == list_end_token {
+                break;
+            }
+            consume_token(tokens, Tok::Comma)?;
+            if allow_trailing_comma && tokens.peek() == list_end_token {
+                break;
+            }
+        }
+    }
+    Ok(v)
```

Comment:
This is probably fine, but I might have written is slightly differently. Consider this version that breaks with a value

```rust
if tokens.peek() != list_end_token { return Ok(vec![]) }

let mut v = vec![];
loop {
    v.push(parse_list_item(tokens)?);
    if tokens.peek() == list_end_token { break Ok(v) }
    consume_token(tokens, Tok::Comma)?;
    if allow_trailing_comma && tokens.peek() == list_end_token { break Ok(v) }
}
```

Reply:
OK. I wrote it that way for the move-lang version (https://github.com/libra/libra/pull/1973). Note that "cargo fmt" puts all the returns and breaks on separate lines so it doesn't look as nice as what you've written here.